dbms/src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -3352,27 +3358,28 @@ bool MergeTreeData::TTLEntry::isPartInDestination(const StoragePolicyPtr & polic
     return false;
 }
 
-const MergeTreeData::TTLEntry * MergeTreeData::selectTTLEntryForTTLInfos(
+std::optional<MergeTreeData::TTLEntry> MergeTreeData::selectTTLEntryForTTLInfos(
         const MergeTreeDataPart::TTLInfos & ttl_infos,
         time_t time_of_move) const
 {
-    const MergeTreeData::TTLEntry * result = nullptr;
-    /// Prefer TTL rule which went into action last.
     time_t max_max_ttl = 0;
+    std::vector<DB::MergeTreeData::TTLEntry>::const_iterator best_entry_it;
 
-    for (const auto & ttl_entry : move_ttl_entries)
+    auto lock = std::lock_guard(move_ttl_entries_mutex);
+    for (auto ttl_entry_it = move_ttl_entries.begin(); ttl_entry_it != move_ttl_entries.end(); ++ttl_entry_it)
     {
-        auto ttl_info_it = ttl_infos.moves_ttl.find(ttl_entry.result_column);
+        auto ttl_info_it = ttl_infos.moves_ttl.find(ttl_entry_it->result_column);
+        /// Prefer TTL rule which went into action last.
         if (ttl_info_it != ttl_infos.moves_ttl.end()
                 && ttl_info_it->second.max <= time_of_move
                 && max_max_ttl <= ttl_info_it->second.max)
         {
-            result = &ttl_entry;
+            best_entry_it = ttl_entry_it;
             max_max_ttl = ttl_info_it->second.max;
         }
     }
 
-    return result;
+    return max_max_ttl ? *best_entry_it : std::optional<MergeTreeData::TTLEntry>();
```

Comment:
`std::optional<MergeTreeData::TTLEntry>();` -- `{}` or `std::nullopt`.

Reply:
First one does not work, thank you for tip!

Reply:
+

Reply:
Nope, both do not work. Because in second operand there is a cast from `TTLEntry` to `optional<TTLEntry>`.